toonz/sources/toonz/toolbar.cpp

```diff
@@ -10,16 +10,121 @@
 #include "toonzqt/menubarcommand.h"
 #include "menubarcommandids.h"
 
+#include "toonz/txshleveltypes.h"
+#include "toonz/txshlevelhandle.h"
+#include "toonz/tframehandle.h"
+#include "toonz/txsheethandle.h"
+#include "toonz/txshcell.h"
+#include "toonz/txshsimplelevel.h"
+#include "toonz/tcolumnhandle.h"
+#include "toonz/preferences.h"
+
 // TnzBase includes
 #include "tenv.h"
 
 #include <QPainter>
 #include <QAction>
 #include <QToolButton>
 #include <QVBoxLayout>
+#include <QObject>
 
 TEnv::IntVar ShowAllToolsToggle("ShowAllToolsToggle", 0);
 
+namespace {
+enum ActivateLevel {
+  Empty       = 0x1,
+  Raster      = 0x2,    //! OVL_XSHLEVEL
+  Scan        = 0x4,    //! TZI_XSHLEVEL
+  ToonzRaster = 0x8,    //! TZP_XSHLEVEL
+  ToonzVector = 0x10,   //! PLI_XSHLEVEL
+  Child       = 0x20,   //! CHILD_XSHLEVEL
+  Mesh        = 0x40,   //! MESH_XSHLEVEL
+  ZeraryFX    = 0x80,   //! ZERARYFX_XSHLEVEL
+  Sound       = 0x100,  //! SND_XSHLEVEL
+  SoundText   = 0x200,  //! SND_TXT_XSHLEVEL
+  Palette     = 0x400,  //! PLT_XSHLEVEL
+
+  Separator = 0x800,
+
+  All = 0x1000,
+};
+
+struct {
+  const char *toolName;
+  bool collapsable;
+  int displayLevels;
+  QAction *action;
+} buttonLayout[] = {
+    {T_Edit, false,
+     (ActivateLevel::Empty | ActivateLevel::Raster | ActivateLevel::Scan |
+      ActivateLevel::ToonzRaster | ActivateLevel::ToonzVector |
+      ActivateLevel::Child | ActivateLevel::Mesh | ActivateLevel::ZeraryFX |
+      ActivateLevel::Palette),
```

Comment:
You are defining here the level types in which each tool is enabled. 
Sorry for the late notice, but can `TTool::getTargetType()` be used instead of this for simplification?
Every tools are defined their target level types on construction and `TTool::getTargetType()` is actually used to enable/disable the tool. ( Please see [`TTool::updateEnabled()`](https://github.com/opentoonz/opentoonz/blob/25d3db05fd57aed51d666455ac8dbc44067a4cea/toonz/sources/tnztools/tool.cpp#L775-L939) )

Reply:
I believe the reason I had chosen not to use this method was because it required constant switching to every tool to determine it's enabled status.  I also did't see an easy way to get a tool without causing it to actually switch to the tool I was checking.

I will look to see what I need to do it this way and make necessary changes to support it.